### PR TITLE
Support wildcard handling in `emit()`

### PR DESCRIPTION
**Pull request by vinibrsl**
Change `emit()` to call handlers registered for parent classes using `isinstance()`. Ensures that base event handlers receive derived events.

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment: Wildcard Support for Event Handling

## Overview
The pull request introduces support for wildcard event handling in the CrewAI event bus system. This enhancement allows handlers to subscribe to all event types, a capability made possible through the use of a new `WILDCARD` constant. The changes affect both the implementation file (`crewai_event_bus.py`) and the corresponding tests (`test_crewai_event_bus.py`).

## Code Quality Findings

### Positive Aspects
- **Clean Implementation**: The wildcard handling is well-implemented, improving flexibility for event subscriptions while maintaining existing functionality.
- **Type Safety**: Good use of type hints, along with Final typing, enhances code clarity and reinforces immutability.
- **Backward Compatibility**: The implementation preserves backward compatibility with existing event handlers, ensuring no disruptions in current usage.
- **Efficiency**: The handler execution logic is streamlined, avoiding unnecessary iterations over handlers.

### Areas for Improvement
1. **Type Safety Improvement**:
   - **Current**: `WILDCARD: Final = CrewEvent`
   - **Recommended**: `WILDCARD: Final[Type[CrewEvent]] = CrewEvent`
   - **Rationale**: Making the type explicit enhances type safety and IDE support, reducing potential runtime errors.

2. **Handler Execution Optimization**:
   - **Current**:
     ```python
     for handler_list in (specific_handlers, wildcard_handlers):
         for handler in handler_list:
             handler(source, event)
     ```
   - **Recommended**:
     ```python
     handlers = specific_handlers + wildcard_handlers
     for handler in handlers:
         try:
             handler(source, event)
         except Exception as e:
             logger.error(f"Error in event handler: {e}")
     ```
   - **Rationale**: This change combines handlers into a single iteration for efficiency and adds error handling for robustness.

3. **Documentation Enhancement**:
   - **Improvement**: Enhance the docstring of the `emit` method to clearly define the role of the `WILDCARD` constant and its impact on handler execution order.
  
## Testing Improvements
- **Current Positive Aspects**:
  - Solid structure and basic coverage in `test_crewai_event_bus.py`, with effective use of pytest fixtures.

- **Suggested Additional Test Cases**:
   - **Multiple Handlers**: Ensure both specific and wildcard handlers operate correctly when registered.
   - **Handler Execution Order**: Validate that specific handlers are executed before wildcard handlers.
   - **Error Handling**: Test that an error within a handler does not disrupt the emission process.
   - **Edge Cases**: Include tests for unregistering handlers and clearing all handlers to validate expected behaviors under these scenarios.

## General Recommendations
1. **Add Logging Support**:
   - Introduce logging for emitted events to facilitate debugging and monitoring, e.g., using the `logging` module to capture the event emission process.

2. **Event Validation**:
   - Incorporate checks to validate the event type to ensure that handlers receive the correct event instance, thereby preventing runtime type errors.

3. **Thread Safety Documentation**:
   - Provide detailed documentation concerning the thread safety of the event bus methods, ensuring that developers are aware of possible concurrency issues.

## Conclusion
This implementation of wildcard support significantly enhances the flexibility of the CrewAI event bus and is primarily well-structured. The proposed improvements focus on increasing robustness, maintainability, and overall developer experience while adhering to best practices. By addressing the recommendations and incorporating the suggested tests, we can ensure that the event bus remains a reliable and adaptable component of the system. 

Thank you for the opportunity to review this PR. I look forward to seeing the enhancements make it into the codebase!

File: src/crewai/utilities/events/crewai_event_bus.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
+from typing import Any, Callable, Dict, Final, List, Type, TypeVar, cast
 
 from blinker import Signal
 
 from crewai.utilities.events.base_events import CrewEvent
 from crewai.utilities.events.event_types import EventTypes
 
 EventT = TypeVar("EventT", bound=CrewEvent)
-
+WILDCARD: Final = CrewEvent(type="wildcard")
```
Comment (vinibrsl) [inline src/crewai/utilities/events/crewai_event_bus.py]: This was to maintain type compatibility with the rest of the code. Suggestions welcome!

File: src/crewai/utilities/events/crewai_event_bus.py [hunk 2]
```diff
@@ -68,9 +68,13 @@ def emit(self, source: Any, event: CrewEvent) -> None:
             event: The event instance to emit
         """
         event_type = type(event)
-        if event_type in self._handlers:
-            for handler in self._handlers[event_type]:
+        specific_handlers = self._handlers.get(event_type, [])
+        wildcard_handlers = self._handlers.get(WILDCARD, [])
```
Comment (bhancockio) [inline src/crewai/utilities/events/crewai_event_bus.py]: Screenshot 2025-03-20 at 8 11 06 AM

I got this error and fixed it by updating the code above. Are you not getting the same warning in your editor?

Screenshot 2025-03-20 at 8 12 26 AM

File: src/crewai/utilities/events/crewai_event_bus.py [hunk 3]
```diff
@@ -68,14 +68,12 @@ def emit(self, source: Any, event: CrewEvent) -> None:
...
-                handler(source, event)
-        self._signal.send(source, event=event)
+        for event_type, handlers in self._handlers.items():
+            if isinstance(event, event_type):
+                for handler in handlers:
+                    handler(source, event)
 
-    def clear_handlers(self) -> None:
-        """Clear all registered event handlers - useful for testing"""
-        self._handlers.clear()
```
Comment (vinibrsl) [inline src/crewai/utilities/events/crewai_event_bus.py:78]: The event bus is a singleton, meaning even if you instantiate a new object, it's reusing the existing one. For that reason, calling this method in tests would make other tests fail. It wasn't being used anywhere else, so I removed it.

File: src/crewai/utilities/events/crewai_event_bus.py [hunk 4]
```diff
@@ -68,14 +68,12 @@ def emit(self, source: Any, event: CrewEvent) -> None:
             event: The event instance to emit
         """
         event_type = type(event)
-        if event_type in self._handlers:
-            for handler in self._handlers[event_type]:
-                handler(source, event)
-        self._signal.send(source, event=event)
+        for event_type, handlers in self._handlers.items():
```
Comment (bhancockio) [inline src/crewai/utilities/events/crewai_event_bus.py:70]: Shouldn't we give `event_type` its own name so it's not overwriting the previous?

**Review by bhancockio — APPROVED** [review]
